blitbuffer.c

```diff
@@ -416,6 +417,65 @@ void BB_fill_rect(BlitBuffer * restrict bb, unsigned int x, unsigned int y, unsi
     }
 }
 
+void BB_fill_rect_color(BlitBuffer * restrict bb, unsigned int x, unsigned int y, unsigned int w, unsigned int h, ColorRGB32 * restrict color) {
+    const int rotation = GET_BB_ROTATION(bb);
+    unsigned int rx, ry, rw, rh;
+    // Compute rotated rectangle coordinates & size
+    switch (rotation) {
+        case 0:
+                rx = x;
+                ry = y;
+                rw = w;
+                rh = h;
+                break;
+        case 1:
+                rx = bb->w - (y + h);
+                ry = x;
+                rw = h;
+                rh = w;
+                break;
+        case 2:
+                rx = bb->w - (x + w);
+                ry = bb->h - (y + h);
+                rw = w;
+                rh = h;
+                break;
+        case 3:
+                rx = y;
+                ry = bb->h - (x + w);
+                rw = h;
+                rh = w;
+                break;
+    }
+
+    // Handle any target pitch properly
+    const int bb_type = GET_BB_TYPE(bb);
+    switch (bb_type) {
+        case TYPE_BBRGB32:
+            // And here either, as we want to preserve the alpha byte
+            if (rx == 0 && rw == bb->w) {
+                // Single step for contiguous scanlines
+                const uint32_t src = (uint32_t) RGBA_To_RGB32(color->r, color->g, color->b, color->alpha);
+                uint32_t * restrict p = (uint32_t *) (bb->data + bb->stride*ry);
+                size_t px_count = bb->pixel_stride*rh;
+                while (px_count--) {
+                    *p++ = src;
+                }
+            } else {
+                // Scanline per scanline
+                const uint32_t src = (uint32_t) RGBA_To_RGB32(color->r, color->g, color->b, color->alpha);
+                for (unsigned int j = ry; j < ry+rh; j++) {
+                    uint32_t * restrict p = (uint32_t *) (bb->data + bb->stride*j) + rx;
+                    size_t px_count = rw;
+                    while (px_count--) {
+                        *p++ = src;
+                    }
+                }
+            }
+            break;
+    }
```

Comment:
Needs RGB565 support

Reply:
For 565 (RGB16) I think we can do ColorRGB32_To_Color16, and then use a uint16_t pointer loop.

For 888 (RGB24) there is no uint24_t type (and this is probably not the time/place to add one).
Given that we can't do `memset(p, v, (rw * 3U));`(setting all three bytes to the same value), I think the best way to approch this is setting the values independently then incrementing the uint8_t pointer by 3 bytes.
```
p[0] = src.r;
p[1] = src.g;
p[2] = src.b;
```
 
 Do the above approaches seem reasonable? I don't have a real device with the above bit layouts to test this on.

Reply:
Or maybe something like memcpy would work since it accepts variable lengths along with any pointer type.
```
memcpy(p, &src, 3);
```